GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,462 @@
+
+	_, err = io.Copy(file, r)
+	return err
+}
+
+func downloadURL(url string, file *os.File) error {
+	// Retry up to 3 times, only wait 1 second between retries.
+	var res *http.Response
+	var err error
+	for i := 1; ; i++ {
+		res, err = http.Get(url)
+		if err != nil && i > 3 {
+			return err
+		}
+		if err == nil {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("GET %q, bad status: %s", url, res.Status)
+	}
+
+	_, err = io.Copy(file, res.Body)
+	return err
+}
+
+func downloadScript(path string, file *os.File) error {
+	ctx := context.Background()
+	bucket, object := findMatch(path)
+	if bucket != "" && object != "" {
+		// Retry up to 3 times, only wait 1 second between retries.
+		for i := 1; ; i++ {
+			err := downloadGSURL(ctx, bucket, object, file)
+			if err == nil {
+				return nil
+			}
+			if err != nil && i > 3 {
+				logger.Infof("Failed to download GCS path: %v", err)
+				break
+			}
+			time.Sleep(1 * time.Second)
+		}
+		logger.Infof("trying unauthenticated download")
+		return downloadURL(fmt.Sprintf("https://%s/%s/%s", storageURL, bucket, object), file)
+	}
+
+	// Fall back to an HTTP GET of the URL.
+	return downloadURL(path, file)
+}
+
+func findMatch(path string) (string, string) {
+	for _, re := range []*regexp.Regexp{gsRegex, gsHTTPRegex1, gsHTTPRegex2, gsHTTPRegex3} {
+		match := re.FindStringSubmatch(path)
+		if len(match) == 3 {
+			return match[1], match[2]
+		}
+	}
+	return "", ""
+}
+
+func runScript(url string, timeout int) int {
+	dir, err := ioutil.TempDir("", "snapshot-scripts")
+	if err != nil {
+		logger.Errorf("failed to create temporary directory")
+	}
+
+	defer os.RemoveAll(dir)
+
+	tmpfile, err := ioutil.TempFile(dir, "snapshot-script")
+	if err != nil {
+		logger.Errorf("failed to create temporary file")
+	}
+
+	if *testMode {
+		data, err := ioutil.ReadFile(url)
+		if err != nil {
+			logger.Errorf("failed to read test script at %s", url)
+		}
+		err = ioutil.WriteFile(tmpfile.Name(), data, 0644)
+		if err != nil {
+			logger.Errorf("failed to copy script at %s to temp file %s", url, tmpfile.Name())
+		}
+	} else {
+		err = downloadScript(url, tmpfile)
+		if err != nil {
+			logger.Errorf("failed to download script at %s", url)
+		}
+	}
+
+	logger.Infof("running script at %s", tmpfile.Name())
+
+	if err := os.Chmod(tmpfile.Name(), 0700); err != nil {
+		logger.Errorf("%v", err)
+	}
+
+	tmpfile.Close()
+
+	cmd := exec.Command(tmpfile.Name())
+	cmd.Start()
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start timer
+	timer := time.After(time.Second * time.Duration(timeout))
+
+	select {
+	case <-timer:
+		logger.Warningf("timeout while running %s", url)
+		// Exit code of the timeout command when it times out
+		return 124
+	case err := <-done:
+		if err != nil {
+			logger.Warningf("sad error %v", err)
+			exiterr := err.(*exec.ExitError)
+			status := exiterr.Sys().(syscall.WaitStatus)
```

Comment:
you're casting it to a type from the syscall package? i don't think that's necessary

Reply:
Hmm I couldn't get it to compile otherwise. I was just following this: https://stackoverflow.com/questions/10385551/get-exit-code-go. Is there a better way?